### ISSUE TITLE
Feature request: onSubscribe property for subscribeToMore for more control (in sync)

### ISSUE THREAD

**Issue by Rusfighter** [apollographql/apollo-client#2863]
Hello there,

Currently I am building a chat application for mobile. However the current state of subscribeToMore does not allow to have some control whenever a (re)subscription is made. This can be especially handy whenever a connection loss happens and a re subscription is made to implement some in sync mechanism.

My proposal: 
SubscribeToMore should have an onSubscribe property which accepts a function. The function can return a object which merges with the variables of the subscription. The onSubscribe function is called whenever the initial subscription is created or at least whenever a re-subscription is made (duo the connection loss & recovery).  This will at least give some opportunities to send a dynamic cursor with the subscription to have some in sync mechanism.

Currently, to achieve this I am hacking the subscription client to catch some events and change the variables: 
```
import { SubscriptionClient } from 'subscriptions-transport-ws';


export default (onSubscriptions = {}) =>
  class extends SubscriptionClient {
    sendMessageRaw(message) {
      if (message.type === 'start' && message.payload && message.payload.query) {
        let additionalVars = {};
        const variables = message.payload.variables;
        const operationName = message.payload.operationName;
        const operationType = message.payload.query.split(` ${message.payload.operationName}`)[0];

        if (operationType === 'subscription' && onSubscriptions[operationName]) {
          additionalVars = onSubscriptions[operationName](variables) || {};
        }

        const payload = {
          ...message.payload,
          variables: {
            ...variables,
            ...additionalVars,
          },
        };

        super.sendMessageRaw({
          ...message,
          payload,
        });
      } else {
        super.sendMessageRaw(message);
      }
    }
  };
```

In summary, currently the subscriptions are not really bi-directional, this will introduce at least some control of how subscriptions should behave (with help of dynamic variables) on the server.

Ilija

**Comment by hwillson** [comment]
To help provide a more clear separation between feature requests / discussions and bugs, and to help clean up the feature request / discussion backlog, Apollo Client feature requests / discussions are now being managed under the https://github.com/apollographql/apollo-feature-requests repository. 

This feature request / discussion will be closed here, but anyone interested in migrating this issue to the new repository (to make sure it stays active), can click [here](https://github.com/apollographql/apollo-feature-requests/issues/new?title=Feature%20request%3A%20onSubscribe%20property%20for%20subscribeToMore%20for%20more%20control%20(in%20sync)&body=Migrated%20from:%20apollographql/apollo-client%232863) to start the migration process. This manual migration process is intended to help identify which of the older feature requests are still considered to be of value to the community. Thanks!
